app/services/xccdf_report_parser.rb

```diff
@@ -33,9 +35,21 @@ def valid_message_format?(message)
     message['id'].present?
   end
 
+  def check_os_version
+    # rubocop:disable Style/GuardClause
+    if benchmark.os_major_version.to_s != @host.os_major_version.to_s
+      raise OSVersionMismatch,
+            "OS major version (#{@host.os_major_version}) does not match with" \
+            " benchmark #{benchmark.ref_id} (#{benchmark.os_major_version})"
+    end
+    # rubocop:enable Style/GuardClause
+  end
+
   def save_all
     Host.transaction do
       save_all_benchmark_info
+      save_all_host_info
+      check_os_version
```

Comment:
We could move this to the top and check `@op_benchmark.id` instead of `@benchmark.ref_id` to fail faster in this situation.

Reply:
Not sure how the `id` would help here, but I'm worried that the `os_major_version` parsing is only available on our `Benchmark` model.

Reply:
`@op_benchmark.id` is actually the benchmark's `ref_id`. It's called `id` in the SCAP data model (kind of confusing, I know). It's a good point that the OS parsing is only in the Benchmark model, but it wouldn't be difficult to make the regex a constant, like `REF_ID_OS_PATTERN = /(?<=RHEL-)\d/` and then use `@op_benchmark.id[Xccdf::Benchmark::REF_ID_OS_PATTERN]` here.

Reply:
Taking the regex out of the models would do tight coupling, which I'd rather avoid. Better way would be to initialize it into `Benchmark`, do the OS check, and then store the benchmark.

Reply:
Works for me

Reply:
It just need to do `save_host` prior to check.